File: src/main/java/xyz/devvydont/smprpg/items/blueprints/block/rawmaterials/RawSilverBlock.java
package xyz.devvydont.smprpg.items.blueprints.block.rawmaterials;

import org.bukkit.NamespacedKey;
import org.bukkit.inventory.CraftingRecipe;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.inventory.recipe.CraftingBookCategory;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.block.CustomBlock;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.blueprints.block.BlockBlueprint;
import xyz.devvydont.smprpg.items.interfaces.ICraftable;
import xyz.devvydont.smprpg.services.ItemService;

import java.util.Collection;
import java.util.List;

public class RawSilverBlock extends BlockBlueprint implements ICraftable {

    public RawSilverBlock(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public CustomBlock getCustomBlock() {
        return CustomBlock.RAW_SILVER_BLOCK;
    }

    @Override
    public NamespacedKey getRecipeKey() {
        return new NamespacedKey(SMPRPG.getInstance(), this.getCustomItemType().getKey() + "_recipe");
    }

    @Override
    public CraftingRecipe getCustomRecipe() {
        var recipe = new ShapedRecipe(this.getRecipeKey(), generate());
        recipe.shape(
                "sss",
                "sss",
                "sss"
        );
        recipe.setIngredient('s', ItemService.generate(CustomItemType.RAW_SILVER));
        recipe.setCategory(CraftingBookCategory.MISC);
        return recipe;
    }

    @Override
    public Collection<ItemStack> unlockedBy() {
        return List.of(
                itemService.getCustomItem(CustomItemType.RAW_SILVER)
        );
    }
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/block/rawmaterials/RawSilverBlock.kt
package xyz.devvydont.smprpg.items.blueprints.block.rawmaterials

import xyz.devvydont.smprpg.block.CustomBlock
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.blueprints.block.BlockBlueprint
import xyz.devvydont.smprpg.items.interfaces.ICompressible
import xyz.devvydont.smprpg.items.interfaces.ICompressible.CompressionStep
import xyz.devvydont.smprpg.items.interfaces.ICustomBlock
import xyz.devvydont.smprpg.services.ItemService

class RawSilverBlock(itemService: ItemService, type: CustomItemType) : BlockBlueprint(itemService, type),
    ICustomBlock, ICompressible {
    override fun getCustomBlock(): CustomBlock {
        return CustomBlock.RAW_SILVER_BLOCK
    }

    override val decompressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.RAW_SILVER_BLOCK ->
            CompressionStep(itemService.getBlueprint(CustomItemType.RAW_SILVER) as ICompressible, 1, 9)
        else -> null
    }

    override val compressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.RAW_SILVER_BLOCK ->
            CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_RAW_SILVER) as ICompressible, 9, 1)
        else -> null
    }
}
